### ISSUE TITLE
Mac Version Crashes Immediately On Launch

### ISSUE THREAD

**Issue by Kaoskitty** [casperstorm/ajour#153]
**Describe the bug**
Installed Ajour and whenever I try to launch it immediately crashes

**To Reproduce**
Try to run Ajour

**Expected behavior**
I would expect the software to launch and be usable

**Software involved**
Please complete the following information:
 - Mac OS 10.15.7
 - Ajour version: 0.3.4

**Additional context**
None

**Log Output**
14:03:55.627 [ajour][DEBUG] Ajour 0.3.4 has started.
14:03:55.661 [ajour::config][DEBUG] loading config
14:03:55.661 [ajour::utility][DEBUG] checking for application update
14:03:55.661 [ajour::theme][DEBUG] loading user themes
14:03:55.662 [ajour::fs::theme][DEBUG] loaded 0 user themes

**Comment by casperstorm** [comment]
@Kaoskitty thanks for reporting this.
I think the issue is related to https://github.com/casperstorm/ajour/pull/138. I fixed it earlier today but it is not released yet.

So to test if this is the case open `Terminal.app` and write `mkdir ~/.config`. After that try to re-launch Ajour.

**Comment by Kaoskitty** [comment]
Hello @casperstorm , 

I checked and I do already have a ~/.config directory which does contain an ajour directory including the files ajour.log, ajour.yaml and themes directory

**Comment by Kaoskitty** [comment]
I was wondering if the could have something to do with the fact that World of Warcraft isn't directly in the Applications folder. I have it in a sub-folder named "Games". Having said that, I do have an alias (symbolic link) to the World of Warcraft folder in Applications as would be expected.

**Comment by tarkah** [comment]
I've improved our logging to capture any program crashes / panics in #157. @Kaoskitty, once we merge this, it'll be helpful to have you rerun so we can understand what is actually crashing the program.

In the meantime, if you're comfortable building this from source, you can `cargo run` and we can see the cause of the panic through that.

**Comment by casperstorm** [comment]
Hi @Kaoskitty
I have built the master branch which have the improved logging incorporated.
Could you try and open this version, and then paste the results from the log?

[ajour-master.dmg.zip](https://github.com/casperstorm/ajour/files/5303725/ajour-master.dmg.zip)

**Comment by Kaoskitty** [comment]
Hi @casperstorm 
I downloaded and tried to run the new version you've linked above, but there seems to actually be less information in the log instead of more. This is all that's in the log now:

08:56:18.641 [ajour][DEBUG] Ajour 0.3.4 has started

**Comment by casperstorm** [comment]
> I also built from source with cargo run and this was the result:
> 
> 09:26:20.193 [ajour][DEBUG] Ajour 0.3.4 has started.
> Illegal instruction

Wasn't there any more information in the console than that? Thank you so much for spending your time trying to solve this.

**Comment by Kaoskitty** [comment]
That was all that was returned in the shell from my cargo run. I did just look in the console at the crash report. I'm guessing you don't want the entire backtrace from the console, but here is the first section:

---------------------------------------------------------------------------------------
Process:               ajour [29603]
Path:                  /Users/USER/*/ajour
Identifier:            ajour
Version:               0
Code Type:             X86-64 (Native)
Parent Process:        tcsh [21194]
Responsible:           iTerm2 [21188]
User ID:               501

Date/Time:             2020-09-30 09:26:20.412 -0700
OS Version:            Mac OS X 10.15.7 (19H2)
Report Version:        12
Anonymous UUID:        8C1B93CD-1D7C-4C27-9385-927361719D79


Time Awake Since Boot: 220000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [29603]

Application Specific Information:
*** CFRelease() called with NULL ***

**Comment by Brightbrown** [comment]
I have been working directly with @casperstorm since I ran into this issue. BUT I am running in 10.12.6 (required because I use a GTX980 GPU which is great for both my Mac and win10 sides... Cupertino refuses to sign any drivers nVidia makes for Mojave and up). Much more debug information goes into the config log file in 0.3.5.

**Comment by casperstorm** [comment]
@Kaoskitty, can you test this version of Ajour?

[ajour-0.3.5-opengl.dmg.zip](https://github.com/casperstorm/ajour/files/5319405/ajour-0.3.5-opengl.dmg.zip)

**Comment by Kaoskitty** [comment]
@casperstorm that version still has the same crashing issue. Here is the output from the ajour.log

12:51:16.134 [ajour][DEBUG] Ajour 0.3.5 has started.
12:51:16.193 [ajour_core::config][DEBUG] loading config
12:51:16.193 [ajour_core::theme][DEBUG] loading user themes
12:51:16.193 [ajour_core::utility][DEBUG] checking for application update
12:51:16.193 [ajour_core::fs::theme][DEBUG] loaded 0 user themes

The crash log looks the same as before. 

Process:               ajour [46200]
Path:                  /Applications/Ajour.app/Contents/MacOS/ajour
Identifier:            com.rogildstorm.ajour
Version:               1.0.0 (1)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           ajour [46200]
User ID:               501

Date/Time:             2020-10-02 12:46:40.181 -0700
OS Version:            Mac OS X 10.15.7 (19H2)
Report Version:        12
Anonymous UUID:        8C1B93CD-1D7C-4C27-9385-927361719D79


Time Awake Since Boot: 400000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   exc handler [46200]

Application Specific Information:
*** CFRelease() called with NULL ***

**Comment by casperstorm** [comment]
Interesting. That means that neither Metal or OpenGL is the issue and that it's something else.
Do you have Xcode installed by any chance with Commandline tools_

**Comment by Kaoskitty** [comment]
Yes, I do have Xcode installed with the command line tools. That's how I built the test version earlier.

**Comment by casperstorm** [comment]
So we agree that you cant get the project to run just by doing `cargo run`?
Can you run it with `RUST_BACKTRACE=1` so I can see everything?

**Comment by Kaoskitty** [comment]
I'm not sure if I'm doing this correctly as I'm not very familiar with rust / cargo build. I'm running the commands:

% git clone https://github.com/casperstorm/ajour.git
% set  set RUST_BACKTRACE=1
% cargo run

And this is all I get back:

     Finished dev [unoptimized + debuginfo] target(s) in 6.20s
     Running `target/debug/ajour`
13:50:59.568 [ajour][DEBUG] Ajour 0.3.4 has started.
Illegal instruction

**Comment by tarkah** [comment]
@Kaoskitty can we get the entire backtrace from the crash report?

**Comment by Kaoskitty** [comment]
Attached is the full backtrace crash report.
[ajour_full_backtrace_log.txt](https://github.com/casperstorm/ajour/files/5321028/ajour_full_backtrace_log.txt)

**Comment by casperstorm** [comment]
Okay, @Kaoskitty.
This seems to be either of the two:

1. A bug in core_graphics and/or font_kit.
2. Or you have broken System Fonts.

Right now 2. should be "easy" to fix. Maybe you can check if you have all system default fonts, or you try to re-download them or something: https://support.apple.com/en-us/HT210192#download

**Comment by Kaoskitty** [comment]
I've taken a look through the fonts and downloaded a few, but there doesn't seem to be any way to check and see if any fonts are broken. My computer does have a completely fresh install of the OS as of about two months ago. I suppose if no one else reports these crashes that it would be more likely to be something on my own computer. I don't know what else I can do to help debug it further.

**Comment by casperstorm** [comment]
I have posted about this in `Iced` our GUI Framework we use: https://github.com/hecrj/iced/issues/548
Let see if we get some valuable information there.

**Comment by casperstorm** [comment]
@Kaoskitty Do you mind trying 0.4.0 opengl: https://github.com/casperstorm/ajour/releases/download/0.4.0/ajour-opengl.dmg?

**Comment by Kaoskitty** [comment]
Same results. Attached is the crash log again, which appears to be pretty much the same.

([ajour_crashLog_8-8-20.txt](https://github.com/casperstorm/ajour/files/5342065/ajour_crashLog_8-8-20.txt))

**Comment by FlyingDiver** [comment]
FWIW, Ajour 0.4.0 ran immediately on my 10.15.7 system.  No issues at all.

**Comment by Kaoskitty** [comment]
I've been trying to do some additional debugging by booting into safe mode and also logging in as a different user. Same results as far as Ajour crashing when trying to run it. However, there is some additional information in the ajour.log file that I haven't seen before. It looks like it's trying to access file in a directory for the user "crs". Not sure if this information helps in the debugging process. Here is the latest complete output from the file ~/.config/ajour/ajour.log

14:14:33.059 [ajour][DEBUG] Ajour 0.4.0 has started.
14:14:33.059 [ajour::gui][DEBUG] antialiasing: false
14:14:34.125 [ajour_core::config][DEBUG] loading config
14:14:34.125 [ajour_core::utility][DEBUG] checking for application update
14:14:34.125 [ajour_core::theme][DEBUG] loading user themes
14:14:34.125 [ajour_core::fs::theme][DEBUG] loaded 0 user themes
14:14:34.777 [panic][ERROR] thread 'main' panicked at 'Open window: NoAvailablePixelFormat': /Users/crs/.asdf/installs/rust/1.46.0/git/checkouts/iced-9e73d2fbe1fce35a/fb015a8/glutin/src/application.rs:64
   0: backtrace::backtrace::trace
   1: backtrace::capture::Backtrace::new
   2: log_panics::init::{{closure}}
   3: std::panicking::rust_panic_with_hook
   4: _rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::result::unwrap_failed
   7: iced_glutin::application::run
   8: ajour::gui::run
   9: ajour::main
  10: std::rt::lang_start::{{closure}}
  11: std::rt::lang_start_internal
  12: _main

**Comment by tarkah** [comment]
The `Open window: NoAvailablePixelFormat` issue might be due to an older OpenGL driver on your machine. Not sure if this command works on Mac or not, but can you run `glxinfo | grep "version"` and paste the output? If not, can you see what OpenGL drivers you have installed?

**Comment by Kaoskitty** [comment]
That command isn't available anymore. OpenGL on Macs seems to have been deprecated and "stuck" at version 4.1 for many years. 

Here is the Apple page regarding what version of OpenGL is installed (I'm running Catalina 10.15.7)

https://support.apple.com/en-us/HT202823

And here is one of several articles I found regarding OpenGL on Macs:

https://apple.stackexchange.com/questions/379326/is-opengl-deprecated-or-removed-in-macos-catalina-10-15

> In Catalina, as in every macOS version going all the way back to Mavericks, the macOS OpenGL implementation is stuck at version 4.1 (2010), and the OpenCL version is stuck at 1.2 (2011). This means that apps that still rely on those APIs on macOS will continue to run, provided they've been updated to meet the 64-bit-only requirement.

**Comment by Brightbrown** [comment]
I just completed an upgrade to High Sierra, 10.13.6 and got the "immediate crash." I had to re-run it several times before I even got a crash report. Had not been an issue in Sierra 10.12.6. After hints that the issue may lie in fonts, I ran Font Book and downloaded all the fonts that listed gray (was very unsure as it looked like almost all were various Asian fonts). BOOM, app is now working as expected. My guess is the app may be trying to access some symbol on one of them.

**Comment by Kaoskitty** [comment]
I've done the same now. Downloaded all of the fonts listed gray in the Font Book, and unfortunately it still crashes the same. I'm running Catalina 10.15.7. 

Which version of Ajour did you download, ajour.dmg or ajour-opengl.dmg? Or did  you compile yourself?

**Comment by Brightbrown** [comment]
Running High Sierra, one thing to note, it very recently had a security update, so I had to get a newer driver from nVidia. No, using the precompiled binaries. Latest 0.5.1 version. The opengl one, the metal one won't run at all, never has.

**Comment by Kaoskitty** [comment]
Interesting. I tried both the metal and OpenGL versions and both crash the same. I was just curious which you tried. Again, I'm running a completely different OS than you, so there may be something else going on too.

**Comment by Brightbrown** [comment]
Even more curious in that under Sierra (at first I got the immediate exit until he created the opengl version), I got all the functionality BUT zero from the graphics... none of the coloring worked so I found a setting that was marginally useful (text washed out best way to describe). At one point he sent me a Unix executable, ran that in Terminal, the message it spit back was it was looking in only one place the the ColorSync framework and that was not where it lived. Without any real qualifications, I think a lot of the issues are in Rust.
